### ISSUE TITLE
Test failure baseservices\\exceptions\\unittests\\StrSwitchFinally\\StrSwitchFinally.cmd

### ISSUE THREAD

**Issue by VincentBu** [dotnet/runtime#49624]
Run: [runtime-coreclr gcstress0x3-gcstress0xc 20210314.1](https://dev.azure.com/dnceng/public/_build/results?buildId=1038574&view=ms.vss-test-web.build-test-results-tab&runId=32191838&resultId=109701&paneView=debug)

Failed test:
```
CoreCLR windows x64 Checked gcstress0xc @ Windows.10.Amd64.Open
 -baseservices\\exceptions\\unittests\\StrSwitchFinally\\StrSwitchFinally.cmd
```

**Error message:**
```
Assert failure(PID 10076 [0x0000275c], Thread: 5064 [0x13c8]): !CREATE_CHECK_STRING(pMT && pMT->Validate())

CORECLR! Object::ValidateInner + 0x14C (0x00007ffcd91ba27c)
CORECLR! Object::Validate + 0x118 (0x00007ffcd91ba0e8)
CORECLR! GcInfoDecoder::ReportRegisterToGC + 0x127 (0x00007ffcd95cafb7)
CORECLR! GcInfoDecoder::ReportSlotToGC + 0xCD (0x00007ffcd95cb1e5)
CORECLR! GcInfoDecoder::EnumerateLiveSlots + 0xCE5 (0x00007ffcd95ca3e9)
CORECLR! EECodeManager::EnumGcRefs + 0x389 (0x00007ffcd90c4899)
CORECLR! GcStackCrawlCallBack + 0x2CC (0x00007ffcd931652c)
CORECLR! Thread::MakeStackwalkerCallback + 0x6B (0x00007ffcd92103f7)
CORECLR! Thread::StackWalkFramesEx + 0x164 (0x00007ffcd9211ca0)
CORECLR! Thread::StackWalkFrames + 0x184 (0x00007ffcd9211a9c)
 File: D:\workspace\_work\1\s\src\coreclr\vm\object.cpp Line: 516
 Image: C:\h\w\A76908BC\p\CoreRun.exe


Return code: 1
Raw output file: C:\h\w\A76908BC\w\B91E0989\e\baseservices\exceptions\Reports\baseservices.exceptions\unittests\StrSwitchFinally\StrSwitchFinally.output.txt
Raw output:
BEGIN EXECUTION
 "C:\h\w\A76908BC\p\corerun.exe" StrSwitchFinally.dll 
-----------------------------
StrSwitchFinalTest
-----------------------------
Expected: 100
Actual: -1073740286
END EXECUTION - FAILED
FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=C:\h\w\A76908BC\p
> C:\h\w\A76908BC\w\B91E0989\e\baseservices\exceptions\unittests\StrSwitchFinally\StrSwitchFinally.cmd
Expected: True
Actual: False

Stack trace
   at baseservices_exceptions._unittests_StrSwitchFinally_StrSwitchFinally_._unittests_StrSwitchFinally_StrSwitchFinally_cmd()

```

**Comment by VincentBu** [comment]
Same test failed with Assert failure(PID 3540 [0x00000dd4], Thread: 5996 [0x176c]): HasPerInstInfo()
Run: [runtime-coreclr gcstress-extra 20210313.1](https://dev.azure.com/dnceng/public/_build/results?buildId=1038123&view=ms.vss-test-web.build-test-results-tab&runId=32183608&resultId=108418&paneView=debug)

Failed test:
```
CoreCLR windows x64 Checked gcstress0xc_zapdisable_heapverify1 @ Windows.10.Amd64.Open
 -baseservices\\exceptions\\unittests\\StrSwitchFinally\\StrSwitchFinally.cmd
```

**Error message:**
```
Assert failure(PID 3540 [0x00000dd4], Thread: 5996 [0x176c]): HasPerInstInfo()

CORECLR! MethodTable::SanityCheck + 0x97 (0x00007ffcb1a4b563)
CORECLR! MethodTable::Validate + 0x17 (0x00007ffcb1a4d24f)
CORECLR! Object::ValidateInner + 0x130 (0x00007ffcb1a5a260)
CORECLR! Object::Validate + 0x118 (0x00007ffcb1a5a0e8)
CORECLR! GcInfoDecoder::ReportRegisterToGC + 0x127 (0x00007ffcb1e6afb7)
CORECLR! GcInfoDecoder::ReportSlotToGC + 0xCD (0x00007ffcb1e6b1e5)
CORECLR! GcInfoDecoder::EnumerateLiveSlots + 0xCE5 (0x00007ffcb1e6a3e9)
CORECLR! EECodeManager::EnumGcRefs + 0x389 (0x00007ffcb1964899)
CORECLR! GcStackCrawlCallBack + 0x2CC (0x00007ffcb1bb652c)
CORECLR! Thread::MakeStackwalkerCallback + 0x6B (0x00007ffcb1ab03f7)
 File: D:\workspace\_work\1\s\src\coreclr\vm\methodtable.h Line: 2821
 Image: C:\h\w\B0BE09A1\p\CoreRun.exe


Return code: 1
Raw output file: C:\h\w\B0BE09A1\w\A7840933\e\baseservices\exceptions\Reports\baseservices.exceptions\unittests\StrSwitchFinally\StrSwitchFinally.output.txt
Raw output:
BEGIN EXECUTION
 "C:\h\w\B0BE09A1\p\corerun.exe" StrSwitchFinally.dll 
-----------------------------
StrSwitchFinalTest
-----------------------------
Expected: 100
Actual: -1073740286
END EXECUTION - FAILED
FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=C:\h\w\B0BE09A1\p
> C:\h\w\B0BE09A1\w\A7840933\e\baseservices\exceptions\unittests\StrSwitchFinally\StrSwitchFinally.cmd
Expected: True
Actual: False

Stack trace
   at baseservices_exceptions._unittests_StrSwitchFinally_StrSwitchFinally_._unittests_StrSwitchFinally_StrSwitchFinally_cmd()

```

**Comment by EgorBo** [comment]
Reproduces for me locally

**Comment by EgorBo** [comment]
Doesn't reproduce if I set `EnableEHWriteThru` to `0`
cc @kunalspathak 

PS: I've wrapped the test with `while (true) {` to reproduce.

**Comment by kunalspathak** [comment]
Thanks, I will take a look.

**Comment by EgorBo** [comment]
@kunalspathak ah, I was testing in my local branch where I messed things up so what I actually reproduced can be a different thin sorry.

**Comment by VincentBu** [comment]
Same test failed in [runtime-coreclr gcstress0x3-gcstress0xc 20210321.1](https://dev.azure.com/dnceng/public/_build/results?buildId=1049735&view=ms.vss-test-web.build-test-results-tab&runId=32444384&resultId=108393&paneView=debug)

Failed test:
```
CoreCLR windows arm64 Checked gcstress0xc @ Windows.10.Arm64v8.Open
 -baseservices\\exceptions\\unittests\\StrSwitchFinally\\StrSwitchFinally.cmd
```

**Error message:**
```
Assert failure(PID 11064 [0x00002b38], Thread: 15272 [0x3ba8]): !CREATE_CHECK_STRING(!"Detected use of a corrupted OBJECTREF. Possible GC hole.")

CORECLR! Object::ValidateInner'::1'::catch$12 + 0x12C (0x00007ff9450aa244)
CORECLR! CallSettingFrame + 0x68 (0x00007ff944ba2f10)
CORECLR! _FrameHandler3::CxxCallCatchBlock + 0x160 (0x00007ff94500b670)
NTDLL! RtlRestoreContext + 0xEC (0x00007ff97502369c)
CORECLR! Object::ValidateInner + 0x14C (0x00007ff944c8ec2c)
CORECLR! Object::Validate + 0xC0 (0x00007ff944c8eab8)
CORECLR! GcInfoDecoder::ReportRegisterToGC + 0x13C (0x00007ff944e586d4)
CORECLR! GcInfoDecoder::ReportSlotToGC + 0xE4 (0x00007ff944e58924)
CORECLR! GcInfoDecoder::EnumerateLiveSlots + 0xD40 (0x00007ff944e57760)
CORECLR! EECodeManager::EnumGcRefs + 0x2CC (0x00007ff944c1d2cc)
 File: D:\workspace\_work\1\s\src\coreclr\vm\object.cpp Line: 593
 Image: D:\h\w\B3FB0995\p\corerun.exe


Return code: 1
Raw output file: D:\h\w\B3FB0995\w\B4200957\e\baseservices\exceptions\Reports\baseservices.exceptions\unittests\StrSwitchFinally\StrSwitchFinally.output.txt
Raw output:
BEGIN EXECUTION
 "D:\h\w\B3FB0995\p\corerun.exe" StrSwitchFinally.dll 
-----------------------------
StrSwitchFinalTest
-----------------------------
Expected: 100
Actual: -1073740286
END EXECUTION - FAILED
FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=D:\h\w\B3FB0995\p
> D:\h\w\B3FB0995\w\B4200957\e\baseservices\exceptions\unittests\StrSwitchFinally\StrSwitchFinally.cmd
Expected: True
Actual: False

Stack trace
   at baseservices_exceptions._unittests_StrSwitchFinally_StrSwitchFinally_._unittests_StrSwitchFinally_StrSwitchFinally_cmd()

```

**Comment by kunalspathak** [comment]
Fixed by #50115
